### DISCUSSION TITLE
Signal processing algorithms

### DISCUSSION THREAD

**Discussion by NoeMurr** [Axect/Peroxide#67] [General]
Hi all,
I noticed the lack of common signal processing algorithms: FFT, convolution, etc. 
Are those algorithm outside the scope of this crate or they have just not been implemented yet?

**Comment by Axect** [comment]
Thank you for your question, @NoeMurr. I appreciate your interest in Peroxide.

Regarding signal processing algorithms like FFT and convolution:

1. At present, we haven't implemented these algorithms directly in Peroxide. There are excellent libraries available, such as RustFFT, which are serving this purpose well for many users.

2. However, this doesn't mean we've ruled out the possibility. If a need arises to implement these algorithms within Peroxide, we would certainly consider doing so.

3. We're always open to community contributions. If someone were to submit a meaningful implementation of these algorithms as a pull request, we would be happy to review and potentially incorporate it into the project.

Peroxide aims to be flexible and responsive to the needs of its users and the community. While we currently rely on existing solutions for certain functionalities, we're always open to expanding our feature set when it benefits our users.

Do you have any specific use cases or requirements for signal processing algorithms within Peroxide that you'd like to see addressed? I'd be interested to hear more about your thoughts on this.

**Reply by NoeMurr** [reply]
Thanks for the answer,
I don't have a specific use case.
The reasons why I was looking into peroxide are multiple:

1. I often work on porting some matlab code implementing some simple dsp algorithms, in C++ and in the last period I used the Armadillo library in C++, that provides a good set of functionalities. 
I'm currently trying to switch technology to Rust, at least in some projects, and I was looking for a good alternative for scientific computing in this language, Peroxide looks very interesting but I saw the lack of those algorithms.

2. I'm also interested in contributing in open source code, so I thought that I could consider to try to implement some of the missing functionalities, but before doing that I needed to check that those implementation would be welcome and not outside the scope of the crate.

**Reply by Axect** [reply]
Thank you for your detailed response. It's great to hear about your background and interests, especially your experience with DSP algorithms and scientific computing.

I want to emphasize that meaningful implementations of numerical algorithms are always welcome in Peroxide. Your experience with porting MATLAB code and using libraries like Armadillo in C++ could be incredibly valuable for our project.

Regarding your interest in contributing:

1. We absolutely welcome contributions for DSP algorithms and other scientific computing functionalities. Your expertise in this area could significantly enhance Peroxide's capabilities.

2. Your contributions would definitely be within the scope of our crate. Peroxide aims to be a comprehensive library for scientific computing in Rust, so implementations of FFT, convolution, and other DSP algorithms would be excellent additions.

3. While we're open to all meaningful contributions, it would be great to maintain some consistency with the existing functions in Peroxide. However, this isn't a strict requirement and is something we can discuss and work on together during the implementation process.

If you decide to contribute, we can collaborate on ensuring that your implementations align well with Peroxide's overall structure and style. This could involve discussions about API design, performance considerations, and integration with existing features.

We're excited about the possibility of your contributions. If you have any specific algorithms in mind or want to discuss implementation details, please feel free to open an issue on our GitHub repository. This would allow us to have a more detailed conversation and potentially involve other community members as well.

Thank you for your interest in improving Peroxide. We look forward to potentially working with you!

**Reply by NoeMurr** [reply]
I'm happy that the feature that I want to develop will be welcome. 

As of now I'm reading and studying the source code to understand the structure and the philosophy behind it's development, 
I will surely open a issue and a subsequent MR when I'll be able to contribute.

Thanks for the information and your time.
